lib/sponsor_users.rb

```diff
@@ -0,0 +1,121 @@
+require_relative 'phone_number'
+require_relative 'email_address'
+
+class SponsorUsers
+  def initialize(user_model:)
+    @user_model = user_model
+  end
+
+  def execute(sponsees, sponsor)
+    sponsor_address = Mail::Address.new(sponsor).address
+
+    return unless EmailAddress.authorised_email_domain?(sponsor_address)
+
+    validate_sponsees(sponsees)
+    invite_sponsees(sponsor, sponsor_address)
+    send_confirmation_email(sponsor_address)
+  end
+
+private
+
+  def invite_sponsees(sponsor, sponsor_address)
+    valid_sponsees.each do |sponsee|
+      if is_valid_email(sponsee)
+        sponsee_address = Mail::Address.new(sponsee).address
+        sponsor_email(sponsor, sponsor_address, sponsee_address)
+      else
+        sponsee_phone_number = PhoneNumber.internationalise_phone_number(sponsee)
+        sponsor_phone_number(sponsor_address, sponsee_phone_number)
+      end
+    end
+  end
+
+  attr_reader :user_model, :valid_sponsees
```

Comment:
It look me a while to figure out what `valid_sponsees` was. Is it randomly hidden between two methods?